frontend/javascripts/dashboard/dataset/queries.tsx

```diff
@@ -0,0 +1,435 @@
+  const queryClient = useQueryClient();
+  const mutationKey = ["folders"];
+
+  return useMutation((folder: FolderUpdater) => updateFolder(folder), {
+    mutationKey,
+    onSuccess: (updatedFolder) => {
+      queryClient.setQueryData(mutationKey, (oldItems: FlatFolderTreeItem[] | undefined) =>
+        (oldItems || []).map((oldFolder: FlatFolderTreeItem) =>
+          oldFolder.id === updatedFolder.id
+            ? {
+                ...updatedFolder,
+                parent: oldFolder.parent,
+              }
+            : oldFolder,
+        ),
+      );
+      queryClient.setQueryData(["folders", updatedFolder.id], updatedFolder);
+    },
+    onError: (err: any) => {
+      handleGenericError(err, "Could not update folder. Check the console for details");
+    },
+  });
+}
+
+export function useMoveFolderMutation() {
+  const queryClient = useQueryClient();
+  const mutationKey = ["folders"];
+
+  return useMutation(
+    ([folderId, newParentId]: [string, string]) => moveFolder(folderId, newParentId),
+    {
+      mutationKey,
+      onMutate: ([folderId, newParentId]) => {
+        // Optimistically update the folder with the new parent.
+        const previousFolders = queryClient.getQueryData(mutationKey);
+        queryClient.setQueryData(mutationKey, (oldItems: FlatFolderTreeItem[] | undefined) =>
+          (oldItems || []).map((oldFolder: FlatFolderTreeItem) =>
+            oldFolder.id === folderId
+              ? {
+                  ...oldFolder,
+                  parent: newParentId,
+                }
+              : oldFolder,
+          ),
+        );
+        return {
+          previousFolders,
+        };
+      },
+      onSuccess: (updatedFolder, [folderId, newParentId]) => {
+        // Use the returned updatedFolder to update the query data
+        queryClient.setQueryData(mutationKey, (oldItems: FlatFolderTreeItem[] | undefined) =>
+          (oldItems || []).map((oldFolder: FlatFolderTreeItem) =>
+            oldFolder.id === updatedFolder.id
+              ? {
+                  ...updatedFolder,
+                  parent: newParentId,
+                }
+              : oldFolder,
+          ),
+        );
+        queryClient.setQueryData(["folders", folderId], () => updatedFolder);
+      },
+      onError: (err: any, _params, context) => {
+        // Restore the old folder tree. Note that without the optimistic update
+        // and the data reversion here, the sidebar would still show the (incorrectly)
+        // moved folder, because the <SortableTree /> component does its own kind of
+        // optimistic mutation by updating its state immediately when dragging a folder.
+        handleGenericError(err, "Could not update folder. Check the console for details");
+
+        if (context) {
+          queryClient.setQueryData(mutationKey, context.previousFolders);
+        }
+      },
+    },
+  );
+}
+
+export function useUpdateDatasetMutation(folderId: string | null) {
```

Comment:
The `update` in the name tripped me up a bit, because I was searching for where the updated values are sent to the server. I think this more of a refresh isn't it?

Reply:
Admittedly, this is not super elegant, but the this mutation can in fact do both (and both are used). I added this comment now:

```js
  /*
   * This mutation can either trigger a simple refresh of the dataset
   * (only pass the APIDatasetId) or it can update the actual dataset
   * when the tuple [APIMaybeUnimportedDataset, string] is passed.
   */
```

It's also used to edit tags, for example.